server/src/lib/mutex.cpp

```diff
@@ -0,0 +1,138 @@
+#include "mutex.h"
+#include <iostream>
+#include <mutex>
+
+namespace fty::storage {
+
+std::optional<fty::storage::Mutex::AccessType> Mutex::m_currentAccess;
+std::atomic<uint16_t>                          Mutex::m_activeReaders;
+std::mutex                                     Mutex::m_mx_currentAccess;
+std::condition_variable                        Mutex::m_cv_access;
+
+Mutex::Mutex()
+{
+}
+
+Mutex::~Mutex()
+{
+}
+
+Mutex::WRITE::WRITE()
+{
+    m = new Mutex();
+    locked = false;
+}
+
+Mutex::WRITE::~WRITE()
+{
+    delete m;
+}
+
+void Mutex::WRITE::lock()
+{
+    if(!locked){
+        locked = true;
+        m->lock(AccessType::WRITE);
+    }
+}
+
+void Mutex::WRITE::unlock()
+{
+    if(locked){
+        locked = false;
+        m->unlock(AccessType::WRITE);
+    }
+}
+
+Mutex::READ::READ()
+{
+    m = new Mutex();
+    locked = false;
+}
+
+Mutex::READ::~READ()
+{
+    delete m;
+}
+
+void Mutex::READ::lock()
+{   
+    if(!locked){
+        locked = true;
+        m->lock(AccessType::READ);
+    }
+}
+
+void Mutex::READ::unlock()
+{   
+    if(locked){
+        locked = false;
+        m->unlock(AccessType::READ);
+    }
+}
+
+Expected<void> Mutex::lock(AccessType access)
+{
+    std::unique_lock<std::mutex> locker(m_mx_currentAccess);
+    if (!m_currentAccess.has_value()) {
+        m_currentAccess = access;
+    } else if (access == AccessType::READ && m_currentAccess != AccessType::READ) {
+        locker.unlock();
+
+        std::mutex                   mx_tmp;
+        std::unique_lock<std::mutex> tmpLocker(mx_tmp);
+
+        m_cv_access.wait(tmpLocker, [&]() {
+            locker.lock();
+            if (m_currentAccess == AccessType::READ || !m_currentAccess.has_value()) {
+                return true;
```

Comment:
leave it here as locked? And this is little bit weird to lock something in condittional variable functor